mailbox/tool/src/main/java/org/apache/james/mailbox/indexer/ReIndexerImpl.java

```diff
@@ -0,0 +1,138 @@
+
+package org.apache.james.mailbox.indexer;
+
+import org.apache.james.mailbox.MailboxManager;
+import org.apache.james.mailbox.MailboxSession;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.indexer.events.FlagsMessageEvent;
+import org.apache.james.mailbox.indexer.events.ImpactingEventType;
+import org.apache.james.mailbox.indexer.events.ImpactingMessageEvent;
+import org.apache.james.mailbox.indexer.registrations.GlobalRegistration;
+import org.apache.james.mailbox.indexer.registrations.MailboxRegistration;
+import org.apache.james.mailbox.model.MailboxPath;
+import org.apache.james.mailbox.model.MessageRange;
+import org.apache.james.mailbox.store.MailboxSessionMapperFactory;
+import org.apache.james.mailbox.store.mail.MessageMapper;
+import org.apache.james.mailbox.store.mail.model.Mailbox;
+import org.apache.james.mailbox.store.mail.model.MailboxId;
+import org.apache.james.mailbox.store.mail.model.Message;
+import org.apache.james.mailbox.store.search.ListeningMessageSearchIndex;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * Note about live re-indexation handling :
+ *
+ *  - Data races may arise... If you modify the stored value between the received event check and the index operation,
+ *  you have an inconsistent behavior.
+ *
+ *  This class is more about supporting changes in real time for future indexed values. If you change a flags / delete
+ *  mails for instance, you will see it in the indexed value !
+ *
+ *  Why only care about updates and deletions ? Additions are already handled by the indexer that behaves normaly. We
+ *  should just "adapt" our indexed value to the latest value, if any. The normal indexer will take care of new stuff.
+ */
+public class ReIndexerImpl<Id extends MailboxId> implements ReIndexer {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ReIndexerImpl.class);
+    public static final int LIMIT = 0;
+
+    private final MailboxManager mailboxManager;
+    private final ListeningMessageSearchIndex<Id> messageSearchIndex;
+    private final MailboxSessionMapperFactory<Id> mailboxSessionMapperFactory;
+
+    public ReIndexerImpl(MailboxManager mailboxManager,
+                         ListeningMessageSearchIndex<Id> messageSearchIndex,
+                         MailboxSessionMapperFactory<Id> mailboxSessionMapperFactory) {
+        this.mailboxManager = mailboxManager;
+        this.messageSearchIndex = messageSearchIndex;
+        this.mailboxSessionMapperFactory = mailboxSessionMapperFactory;
+    }
+
+    public void reIndex(MailboxPath path) throws MailboxException {
+        MailboxSession mailboxSession = mailboxManager.createSystemSession("re-indexing", LOGGER);
+        reIndex(path, mailboxSession);
+    }
+
+
+    public synchronized void reIndex() throws MailboxException {
```

Comment:
what do you think you are protecting with this synchronized ? why no synchronized when you index a single mailbox ? what do you expect in distributed context ?